crypto/provider_core.c

```diff
@@ -418,8 +418,7 @@ OSSL_PROVIDER *ossl_provider_find(OSSL_LIB_CTX *libctx, const char *name,
          * them.
          */
         if (!noconfig) {
-            if (ossl_lib_ctx_is_default(libctx))
-                OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
+            OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
```

Comment:
This change seems to force the config file to be loaded into the default libctx, when we're trying to find a provider in a non-default libctx. Why?

Reply:
Basically we always want to do the INIT_LOAD_CONFIG otherwise it will be called during the cleanup anyway.

Reply:
> otherwise it will be called during the cleanup anyway

What is causing it to be called during cleanup?

Reply:
I suppose the cleanup of the default context calls ossl_provider_find. Or something like that.

Reply:
So can we somehow suppress the loading of config during cleanup? E.g. `ossl_provider_find` has an argument which lets you specify whether config should be automatically loaded or not.

Reply:
IMO it actually makes sense to do the init when loading a provider into non-default contexts as well. I am not saying the cleanup cannot be improved but this change IMO should stay.

Reply:
The loading of the configuration file of course affects non-default contexts as well. So the condition in the current code does not make sense.

Reply:
> The loading of the configuration file of course affects non-default contexts as well

While this technically *could* be true (depending on what the application does) it is not actually relevant. We absolutely want to minimise the places where we load the config file. Loading the config file is a heavy operation that holds a lock. We have had numerous problems in the past with indiscriminate loading of the config file all over the place. The more places this occurs the more likely we are to run into deadlock situations. This has happened previously and it was a pain to sort out.

In order to avoid this we load the config file on a just-in-time basis *only* in the places where we are about to read some variable that might be affected by the config file (in this case the `store->providers` stack). We might well need the config file loaded for *other things* but not for the thing that we are about to do. When we get around to doing those other things *that* is the correct time to attempt to load the config file.

I am not keen on this change. It is papering over a different problem. Namely that we attempt to load the config file during cleanup, which does not make sense. It seems to fix the problem merely by chance, i.e. perhaps there are other routes that might end up incorrectly attempting to load the config file during cleanup which don't involve a non-default libctx.

Reply:
I do not understand this so - we are loading the config file here anyway if the context is default. But individual provider configuration does not depend on whether the context is default or not. So either we should not be loading the config file in this function at all or we should be loading it always. The `if()` does not make sense.

Reply:
This is about *automatic* loading of the config file. Automatic loading of the config file only ever impacts the default libctx. It has no impact on non-default libctxs. You can *explicitly* load a config file into a non-default libctx using `OSSL_LIB_CTX_load_config`, but that is not relevant to this code.

Reply:
Hmm, this is true. Except I see some strange behavior with PBKDF2.